homeassistant/components/weatherflow/sensor.py

```diff
@@ -0,0 +1,383 @@
+) -> None:
+    """Set up WeatherFlow sensors using config entry."""
+
+    @callback
+    def async_add_sensor(device: WeatherFlowDevice) -> None:
+        """Add WeatherFlow sensor."""
+        LOGGER.debug("Adding sensors for %s", device)
+
+        sensors: list[WeatherFlowSensorEntity | WeatherFlowAirDensitySensorEntity] = [
+            WeatherFlowSensorEntity(
+                device=device,
+                description=description,
+                is_metric=(hass.config.units == METRIC_SYSTEM),
+            )
+            for description in SENSORS
+            if hasattr(device, description.key)
+        ]
+
+        sensors = sensors + [
+            WeatherFlowAirDensitySensorEntity(
+                device=device,
+                description=description,
+                is_metric=(hass.config.units == METRIC_SYSTEM),
+            )
+            for description in VOC_SENSORS
+            if hasattr(device, description.key)
+        ]
+
+        async_add_entities(sensors)
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            f"{DOMAIN}_{config_entry.entry_id}_add_{SENSOR_DOMAIN}",
+            async_add_sensor,
+        )
+    )
+
+
+class WeatherFlowSensorEntity(SensorEntity):
+    """Defines a WeatherFlow sensor entity."""
+
+    entity_description: WeatherFlowSensorEntityDescription
+    _attr_should_poll = False
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        device: WeatherFlowDevice,
+        description: WeatherFlowSensorEntityDescription,
+        is_metric: bool = True,
+    ) -> None:
+        """Initialize a WeatherFlow sensor entity."""
+        self.device = device
+        self.entity_description = description
+
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, device.serial_number)},
+            manufacturer="WeatherFlow",
+            model=device.model,
+            name=f"{device.model} {device.serial_number}",
+            sw_version=device.firmware_revision,
+        )
+
+        self._attr_unique_id = f"{device.serial_number}_{description.key}"
+
+        # In the case of the USA - we may want to have a suggested US unit which differs from the internal suggested units
+        if (description.imperial_suggested_unit is not None) and (not is_metric):
+            self._attr_suggested_unit_of_measurement = (
+                description.imperial_suggested_unit
+            )
+
+    @property
+    def last_reset(self) -> datetime | None:
+        """Return the time when the sensor was last reset, if any."""
+        if self.entity_description.state_class == SensorStateClass.TOTAL:
+            return self.device.last_report
+        return None
+
+    @property
+    def native_value(self) -> datetime | StateType:
+        """Return the state of the sensor."""
+
+        # Extract & process raw sensor data
+        raw_sensor_data = getattr(self.device, self.entity_description.key)
```

Comment:
You dont mean this do you?

```python

@dataclass
class WeatherFlowSensorEntityDescription(SensorEntityDescription):
    """Describes a WeatherFlow sensor entity description."""

    event_subscriptions: list[str] = field(default_factory=lambda: [EVENT_OBSERVATION])
    imperial_suggested_unit: None | str = None


    def get_native_value(self, device) -> datetime | StateType:
        """Return the parsed sensor value."""
        raw_sensor_data = getattr(device, self.key)
        LOGGER.debug(f"{self.key} == {type(raw_sensor_data)}")
        if isinstance(raw_sensor_data, Quantity):
            sensor_value = raw_sensor_data.magnitude
            return sensor_value
        if isinstance(raw_sensor_data, Enum):
            sensor_value = raw_sensor_data.name.lower()
            return sensor_value
        if isinstance(raw_sensor_data, float):
            return raw_sensor_data
        if isinstance(raw_sensor_data, int):
            return raw_sensor_data
        return None
```

Reply:
Or i can write a lambda for the raw_data conversion.. and then thats probably what you arelooking for:


```
    raw_data_conv_fn= lambda raw_data: raw_data.magnitude
    raw_data_conv_fn= lambda raw_data: raw_data
    raw_data_conv_fn= lambda raw_data: raw_data.name.lower()
```

Reply:
Not completely, he means a 
```python
SENSORS: tuple[RDWSensorEntityDescription, ...] = (
    RDWSensorEntityDescription(
        key="apk_expiration",
        translation_key="apk_expiration",
        device_class=SensorDeviceClass.DATE,
        value_fn=lambda vehicle: vehicle.apk_expiration,
    ),
    RDWSensorEntityDescription(
        key="ascription_date",
        translation_key="ascription_date",
        device_class=SensorDeviceClass.DATE,
        value_fn=lambda vehicle: vehicle.ascription_date,
    ),
)
```
You create a `value_fn` (or whatever name youchoose, but value_fn is a common name), and that function returns what you want.

You already know what sensors return an enum and which ones return a Quantity, so what you do is:
```python
    WeatherFlowSensorEntityDescription(
        key="dew_point_temperature",
        translation_key="dew_point",
        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
        device_class=SensorDeviceClass.TEMPERATURE,
        state_class=SensorStateClass.MEASUREMENT,
        suggested_display_precision=1,
        value_fn=lambda device: device.dew_point_temperature.magnitude
    ),
```
Probably not the right sensor, but you get the point

Reply:
This way you keep the logic on what data to use as value in the same place as the sensor definition, making it easier to get what the code (And that sensor) does. It also makes easier to change one single sensor that is different from the rest.

If the amount of Quantity and enum sensors are that large, you an always create a helper function to do the .lower() and .magnitude for you, but iirc you had like 2 enums, and idk how much quantity so I wouldn't necessarily advice this

Reply:
Most sensors use magnitude so that would be the default - and then some like this:

```python
    WeatherFlowSensorEntityDescription(
        key="precipitation_type",
        translation_key="precipitation_type",
        device_class=SensorDeviceClass.ENUM,
        options=["none", "rain", "hail", "rain_hail", "unknown"],
        icon="mdi:weather-rainy",
        raw_data_conv_fn= lambda raw_data: raw_data.name.lower()
    ),
```

get custom conversions?

Reply:
Anyways ... off to the gym... i'll get back to this later today ..

Reply:
(just need to figure out the typing)..